### PR TITLE
Chart: add support for container lifecycle hooks

### LINKED ISSUE #32402: [Chart] Add support for container lifecycle
### Description

Adding possibility to configure [lifecycle](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/) to on  service/node level.

### Use case/motivation

Currently configuring [lifecycle](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/) to containers in  service/node level is not possible. There are some use cases which in which PostStart and PreStop could be utilized. For example  new templated value can be configured on global and service/node level not dissimilar to the new [securityContexts](https://github.com/apache/airflow/pull/31043).

### Related issues

https://github.com/apache/airflow/issues/27612

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by rzjfr**
This PR attempts to add possibility to add [lifecycle](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/) to containers. There are some use cases which in which `PostStart` and `PreStop` could be utilized. The new `containerLifecycleHooks` (templated) can be configured on global and service/node level not dissimilar to the new `securityContexts` variable.  

closes: https://github.com/apache/airflow/issues/32402

### CODE REVIEW

File: chart/files/pod-template-file.kubernetes-helm-yaml [hunk 1]
```diff
@@ -23,6 +23,7 @@
 {{- $topologySpreadConstraints := or .Values.workers.topologySpreadConstraints .Values.topologySpreadConstraints }}
 {{- $securityContext := include "airflowPodSecurityContext" (list . .Values.workers) }}
 {{- $containerSecurityContext := include "containerSecurityContext" (list . .Values.workers) }}
+{{- $containerLifecycleHooks := include "containerLifecycleHooks" (list . .Values.workers) }}
```
Comment (jedcunningham) [inline chart/files/pod-template-file.kubernetes-helm-yaml]: I'm not sure we need a named template for this. We can do this:

```suggestion
{{- $containerLifecycleHooks := or .Values.workers.containerLifecycleHooks .Values.ContainerLifecycleHooks }}
```

File: chart/files/pod-template-file.kubernetes-helm-yaml [hunk 2]
```diff
@@ -64,6 +65,9 @@ spec:
       image: {{ template "pod_template_image" . }}
       imagePullPolicy: {{ .Values.images.pod_template.pullPolicy }}
       securityContext: {{ $containerSecurityContext | nindent 8 }}
+      {{- if $containerLifecycleHooks }}
+      lifecycle: {{- $containerLifecycleHooks | nindent 8 }}
```
Comment (jedcunningham) [inline chart/files/pod-template-file.kubernetes-helm-yaml]: And to the tpl here:

```suggestion
      lifecycle: {{- tpl $containerLifecycleHooks . | nindent 8 }}
```

File: chart/values.yaml [hunk 3]
```diff
@@ -43,6 +43,11 @@ securityContexts:
   pod: {}
   containers: {}
 
+# Default container level lifecycle hooks for every container
+# See:
+# https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/
```
Comment (jedcunningham) [inline chart/values.yaml]: I don't think we need to link folks to this everywhere. Easy enough to search for it.

```suggestion
```

File: helm_tests/airflow_aux/test_airflow_common.py [hunk 4]
```diff
@@ -408,3 +408,39 @@ def test_priority_class_name(self):
...
+                "templates/workers/worker-deployment.yaml",
+                "templates/jobs/create-user-job.yaml",
+                "templates/jobs/migrate-database-job.yaml",
+                "templates/triggerer/triggerer-deployment.yaml",
+                "templates/statsd/statsd-deployment.yaml",
+                "templates/redis/redis-statefulset.yaml",
+            ],
+        )
+
+        for index in range(len(docs) - 2):
```
Comment (jedcunningham) [inline helm_tests/airflow_aux/test_airflow_common.py]: Can you refactor this to use list slices instead? e.g like [this](https://github.com/apache/airflow/blob/0fbb05751acf6061f9f6fa83ee53816d568055d0/helm_tests/security/test_security_context.py#L247).

File: helm_tests/airflow_aux/test_airflow_common.py [hunk 5]
```diff
@@ -408,3 +408,39 @@ def test_priority_class_name(self):
             priority = doc["spec"]["template"]["spec"]["priorityClassName"]
 
             assert priority == f"low-priority-{component}"
+
+    # Test containerLifecycleHooks for main containers
+    def test_global_container_lifecycle_webhooks(self):
+        post_start_value = {"exec": {"command": ["bash", "-c", "echo postStart"]}}
```
Comment (jedcunningham) [inline helm_tests/airflow_aux/test_airflow_common.py]: Can you add something that exercises templating?

**Comment by rzjfr** [comment]
@jedcunningham Thank you for the review, good points, I should have applied all of the suggested changes on the last commit,
I also consolidated the tests in a single file, if it the structure is fine I can try adding more test cases.

**Review by potiuk — APPROVED** [review]
LGTM

**Comment by potiuk** [comment]
Rebased to see if everything works fine with latest main. @jedcunningham ?

File: helm_tests/airflow_aux/test_container_lifecycle.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+# under the License.
+from __future__ import annotations
+
+import jmespath
+import pytest
+
+from tests.charts.helm_template_generator import render_chart
+
+
+@pytest.mark.parametrize(
```
Comment (jedcunningham) [inline helm_tests/airflow_aux/test_container_lifecycle.py]: Can we not parametrize the whole class, just the places that really require it? These tests are slow enough as it is :)

I think most of these could just get away with using constants instead.

File: helm_tests/airflow_aux/test_pod_template_file.py [hunk 7]
```diff
@@ -750,3 +750,34 @@ def test_workers_priority_class_name(self):
...
         assert "test-priority" == jmespath.search("spec.priorityClassName", docs[0])
+
+    @pytest.mark.parametrize(
+        "release_name,lifecycle_tempated,lifecycle_parsed,hook_type",
+        [
+            (
+                "test-release",
+                {"exec": {"command": ["echo", "postStart", "{{ .Release.Name }}"]}},
+                {"exec": {"command": ["echo", "postStart", "test-release"]}},
+                "postStart",
```
Comment (jedcunningham) [inline helm_tests/airflow_aux/test_pod_template_file.py]: Doing 1, preStop or postStart, should be sufficient.

**Comment by rzjfr** [comment]
@jedcunningham thanks again for the review, I should have applied both of the suggested changes on my last commit, please let me know in case I missed something.

**Review by jedcunningham — APPROVED** [review]

**Comment by jedcunningham** [comment]
Thanks @rzjfr! Congrats on your first commit 🎉

**Pull request merged**